### PR TITLE
gh-103906: Remove immortal refcounting in compile/marshal.c

### PR DESCRIPTION
**Pull request by corona10**
* Issue: gh-103906

### CODE REVIEW

**Comment by corona10** [comment]
@brandtbucher Not sure worth to do it at compile.c as much as an interpreter but it looks possible.

File: Python/compile.c [hunk 1]
```diff
@@ -892,10 +892,10 @@ static PyObject*
 merge_consts_recursive(PyObject *const_cache, PyObject *o)
 {
     assert(PyDict_CheckExact(const_cache));
-    // None and Ellipsis are singleton, and key is the singleton.
+    // None and Ellipsis are immortal object, and key is the singleton.
```
Comment (JelleZijlstra) [inline Python/compile.c]:
```suggestion
    // None and Ellipsis are immortal objects, and key is the singleton.
```

**Review by JelleZijlstra — COMMENTED** [review]
If you grep for `Py_NewRef(Py_Ellipsis)` there are a few more in other files, though I doubt it will make a measurable difference in performance.

File: Python/compile.c (hunk 1 above)
Comment (corona10) [reply Python/compile.c]: > though I doubt it will make a measurable difference in performance.

Yeah, I think so too.

**Comment by iritkatriel** [comment]
I'm a bit concerned that a change like this would make it hard to reverse the decision to make these objects immortal in the future (it is no longer clear where we need a new reference and where we don't).

Should we have Py_New_None/Py_New_Elipsis/etc macros instead, so that the code is clear about this?

**Review by JelleZijlstra — APPROVED** [review]

**Comment by corona10** [comment]
> Should we have Py_New_None/Py_New_Elipsis/etc macros instead, so that the code is clear about this?

I think that it will be the proper way and easy to track.

**Comment by JelleZijlstra** [comment]
> Should we have Py_New_None/Py_New_Elipsis/etc macros instead, so that the code is clear about this?

I'm not sure that will be workable. Since there is no refcount change, there will be no way to enforce that these macros are used correctly, so we're likely to regress very quickly on correct use of those macros.

**Comment by corona10** [comment]
@vstinner What do you think about this agenda?

**Comment by iritkatriel** [comment]
> > Should we have Py_New_None/Py_New_Elipsis/etc macros instead, so that the code is clear about this?
> 
> I'm not sure that will be workable. Since there is no refcount change, there will be no way to enforce that these macros are used correctly, so we're likely to regress very quickly on correct use of those macros.

True. We would need a test mode where they are not immortal and the macros actually change the refcount.

**Comment by corona10** [comment]
> True. We would need a test mode where they are not immortal and the macros actually change the refcount.

I am not sure that I understand your intention correctly, so do we need to introduce the following configuration things?
`./configure --emulate-ref-count`
So if such a configuration is enabled, Py_New_None will do ref counting?

**Comment by corona10** [comment]
One of my concerns about adding such a configuration which determining the C API or macro behavior, as @JelleZijlstra mentioned it should always be run under at least from the build bot with the configuration enabled. Without the integration from the CI level, such configuration is easily broken and can not guarantee to be built successfully at any time.

This is one of my lessons from the `--with-experimental-isolated-subinterpreters` flag:
https://github.com/python/cpython/issues/90988

**Comment by corona10** [comment]
My suggestion:
How about adding test cases for immortal objects from the C API side? 
If we add the test code that can occur weird refcount, under the immortal object condition, the test will be passed but if we disable immortal object condition from specific object, the test will be failed and we can easily track what should be fixed.

**Comment by iritkatriel** [comment]
We need a decision on whether we want to put some effort into making the move immortality reversible, or not. The technicalities are not very difficult once that decision is made.  I'm not sure who should make the call, but it's not me.

**Comment by corona10** [comment]
It looks like https://github.com/python/cpython/pull/105195 will be merged soon.
Only applying this PR for Python 3.13 will not be harmful if the documentation exists.

I merge this PR, please let me know if the PR occurs any problem.

**Pull request merged**